codegen/src/getable.rs

```diff
@@ -0,0 +1,223 @@
+
+        // data contains the the data for each field of a variant; the variant of the passed value
+        // is defined by the tag(), which is defined by order of the variants (the first variant is 0)
+        cons = quote! {
+            match data.tag() {
+                #(#variants,)*
+                tag => panic!("Unexpected tag: '{}'. Do the type definitions match?", tag)
+            }
+        };
+    }
+
+    gen_impl(ident, generics, cons)
+}
+
+fn gen_impl(ident: Ident, generics: Generics, cons_expr: TokenStream) -> TokenStream {
+    // add trailing comma or add a where if there are no predicates
+    let (_, ty_generics, where_clause) = generics.split_for_impl();
+    let where_clause = where_clause
+        .map(|clause| quote! { #clause, })
+        .unwrap_or(quote! { where });
+
+    // lifetime bounds like '__vm: 'a, 'a: '__vm (which implies => 'a == '__vm)
+    // writing bounds like this is a lot easier than actually replacing all lifetimes
+    // with '__vm
+    let lifetime_bounds = create_lifetime_bounds(&generics);
+
+    // generate bounds like T: Getable for every type parameter
+    let getable_bounds = create_getable_bounds(&generics);
+
+    // generate the generic params for the impl block
+    // these need an additional lifetime used by Getable
+    let mut generics = generics.clone();
+    let lifetime = GenericParam::from(LifetimeDef::new(Lifetime::new("'__vm", Span::call_site())));
+    generics.params.insert(0, lifetime);
+    let (impl_generics, ..) = generics.split_for_impl();
+
+    quote! {
+        #[automatically_derived]
+        #[allow(unused_attributes)]
+        impl #impl_generics ::gluon::vm::api::Getable<'__vm> for #ident #ty_generics
+        #where_clause #(#getable_bounds,)* #(#lifetime_bounds),*
+        {
+            fn from_value(vm: &'__vm ::gluon::vm::thread::Thread, variants: ::gluon::vm::Variants) -> Self {
+                let data = match variants.as_ref() {
+                    ::gluon::vm::api::ValueRef::Data(data) => data,
+                    val => panic!("Unexpected value: '{:?}'. Do the type definitions match?", val),
+                };
+
+                #cons_expr
+            }
+        }
+    }
+}
+
+fn gen_variant_match(ident: &Ident, tag: VmTag, variant: &Variant) -> TokenStream {
+    let variant_ident = &variant.ident;
+
+    // depending on the type of the variant we need to generate different constructors
+    // for the enum
+    match &variant.fields {
+        Fields::Unit => quote! {
+            #tag => #ident::#variant_ident
+        },
+        // both constructors that need to marshall values extract them by using the index
+        // of the field to get the content from Data::get_variant;
+        // the data variable was assigned in the function body above
+        Fields::Unnamed(FieldsUnnamed { unnamed, .. }) => {
+            let cons = gen_tuple_cons(unnamed.into_iter().collect());
```

Comment:
Could do without the collect here if the function were to take an iterator or the original `unnamed` slice